### PR TITLE
SI-8731 warning if @switch is ignored

### PR DESCRIPTION
**Pull request by lrytz**
For matches with two or fewer cases, @switch is ignored. This should
not happen silently.

### CODE REVIEW

**Comment by lrytz** [comment]
Review by @gkossakowski

File: src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala [hunk 1]
```diff
@@ -550,8 +550,20 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
...
+                case a :: b :: c :: _ => 3
+                case cases =>
+                  cases.map({
+                    case AlternativesTreeMaker(_, alts, _) :: _ => lengthMax3(alts)
+                    case c => 1
+                  }).sum
+              }
+              val requireSwitch = hasSwitchAnnotation && lengthMax3(casesNoSubstOnly) > 2
+              if (hasSwitchAnnotation && !requireSwitch)
+                reporter.warning(scrut.pos, "ignoring @switch, warnings are only issued for matches with more than two cases")
```
Comment (jozic) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala]: shouldn't it be `for matches with two or less cases`?

Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala]: no, but maybe i have to come up with a clearer wording.

Comment (jozic) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala]: maybe :)
i understood this as if you have a pm with 1 | 2 cases and switch annotation, than this warning is issued, as for 2 or less cases switch is ignored.

**Comment by lrytz** [comment]
cc @adriaanm

File: src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala [hunk 2]
```diff
@@ -550,8 +550,20 @@ trait MatchTreeMaking extends MatchCodeGen with Debugging {
                 case _ => false
               }
               val suppression = Suppression(suppressExhaustive, supressUnreachable)
+              val hasSwitchAnnotation = treeInfo.isSwitchAnnotation(tpt.tpe)
               // matches with two or fewer cases need not apply for switchiness (if-then-else will do)
-              val requireSwitch = treeInfo.isSwitchAnnotation(tpt.tpe) && casesNoSubstOnly.lengthCompare(2) > 0
+              // `case 1 | 2` is considered as two cases.
+              def lengthMax3[T](l: List[T]): Int = l match {
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala]: Maybe wrap the `lengthMax3(casesNoSubstOnly) > 2` in  `def exceedsTwoCasesOrAlts` (to explain what kind of 'length' we're measuring), which contains `lengthMax3` as a local method?

Comment (lrytz) [inline src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala]: sure, thanks. so in general you agree with the additional warning?

**Comment by gkossakowski** [comment]
I think the warning makes sense to introduce. I agree with @adriaanm's suggestion to clarify the condition as expressed in code.

**Comment by lrytz** [comment]
OK. I'll get it fixed up later today.

**Comment by lrytz** [comment]
created a method `exceedsTwoCasesOrAlts` and also changed the warning message. now it says `matches with two cases or fewer are emitted using if-then-else instead of switch`.

**Comment by adriaanm** [comment]
LGTM

**Pull request merged**